File: events/2024/04/04.py
def preprocessing(puzzle_input):
    letters = {'X': set(), 'M': set(), 'A': set(), 'S': set()}
    for y, line in enumerate(puzzle_input.splitlines()):
        for x, c in enumerate(line):
            letters[c].add((x, y))
    return letters

def solver(letters):
    yield find_xmas(letters)
    yield find_x_mas(letters)
    
def find_xmas(letters):
    cnt = 0
    for (x, y) in letters['X']:
        if (x - 1, y+1) in letters['M'] and (x - 2, y+2) in letters['A'] and (x - 3, y+3) in letters['S']: cnt += 1
        if (x - 1, y) in letters['M'] and (x - 2, y) in letters['A'] and (x - 3, y) in letters['S']: cnt += 1 # a gauche 
        if (x - 1, y-1) in letters['M'] and (x - 2, y-2) in letters['A'] and (x -3, y-3) in letters['S']: cnt += 1 #
        if (x, y+1) in letters['M'] and (x, y+2) in letters['A'] and (x, y+3) in letters['S']: cnt += 1
        if (x, y-1) in letters['M'] and (x, y-2) in letters['A'] and (x, y-3) in letters['S']: cnt += 1
        if (x + 1, y+1) in letters['M'] and (x + 2, y+2) in letters['A'] and (x + 3, y+3) in letters['S']: cnt += 1 # en bas à droite
        if (x + 1, y) in letters['M'] and (x + 2, y) in letters['A'] and (x + 3, y) in letters['S']: cnt += 1 # a droite
        if (x + 1, y-1) in letters['M'] and (x + 2, y-2) in letters['A'] and (x + 3, y-3) in letters['S']: cnt += 1
    return cnt

def find_x_mas(letters):
    cnt = 0
    for (x, y) in letters['A']:
        if (x-1, y-1) in letters['M'] and (x+1, y-1) in letters['M'] and (x+1, y+1) in letters['S'] and (x-1, y+1) in letters['S']: cnt +=1
        elif (x-1, y-1) in letters['M'] and (x-1, y+1) in letters['M'] and (x+1, y-1) in letters['S'] and (x+1, y+1) in letters['S']: cnt +=1
        elif (x+1, y-1) in letters['M'] and (x+1, y+1) in letters['M'] and (x-1, y-1) in letters['S'] and (x-1, y+1) in letters['S']: cnt +=1
        elif (x-1, y+1) in letters['M'] and (x+1, y+1) in letters['M'] and (x+1, y-1) in letters['S'] and (x-1, y-1) in letters['S']: cnt +=1
    return cnt

File: events/2024/04/04.rs
aoc::main!();


//     Preprocessing consists in going through the puzzle input 
//     and stock position of the letters 'X', 'M', 'A' and 'S'.
fn preprocessing(puzzle_input: &str) ->  HashMap<char,HashSet<(isize, isize)>> {
    let mut letters = HashMap::from([
                       ('X', HashSet::new()), ('M', HashSet::new()),
                       ('A', HashSet::new()), ('S', HashSet::new())]);
    puzzle_input
    .lines()
    .enumerate()
    .for_each(|(y, line)| {
        line.char_indices().for_each(|(x, c)| {
            let letter = letters.get_mut(&c).unwrap();
            letter.insert((x as isize, y as isize));
        });
    });
    letters
}


//     Patterns to be detected are:
//     XMAS    X...    X...    ...X    SAMX    S...    S...    ...S
//     ....    .M..    M...    ..M.    ....    .A..    A...    ..A.
//     ....    ..A.    A...    .A..    ....    ..M.    M...    .M..
//     ....    ...S    S...    S...    ....    ...X    X...    X...
//     This function starts from every detected 'X' then check if a 
//     "MAS" is formed in any of the possible directions
fn part_1(letters: HashMap<char,HashSet<(isize, isize)>>) -> usize {
    letters.get(&'X').unwrap()
    .iter()
    .map(|&(x, y)| 
        [(1, 0), (1, 1), (0, 1), (-1, 1), (-1, 0), (-1, -1), (0, -1), (1, -1)]
        .iter()
        .filter(|(dx, dy)|
            letters.get(&'M').unwrap().contains(&(x +     dx, y +     dy)) &&
            letters.get(&'A').unwrap().contains(&(x + 2 * dx, y + 2 * dy)) &&
            letters.get(&'S').unwrap().contains(&(x + 3 * dx, y + 3 * dy)))
        .count())
    .sum()
}


//     Patterns to be detected are:
//     M.S     S.M     M.M     S.S
//     .A.     .A.     .A.     .A.
//     M.S     S.M     S.S     M.M
//     
//     Readind left-to-right, top-to-bottom without considering the 'A', 
//     the patterns are "MSMS", "SMSM", "MMSS" and "SSMM".
//     This function therefore look for this patterns around all 'A'. 
fn part_2(letters: HashMap<char,HashSet<(isize, isize)>>) -> usize {
    letters.get(&'A').unwrap()
    .iter()
    .filter(|&(x, y)| 
        [('M','S','M','S'), ('S','M','S','M'), ('M','M','S','S'), ('S','S','M','M')]
        .iter()
        .any(|l|
            letters.get(&l.0).unwrap().contains(&(x - 1, y - 1)) &&
            letters.get(&l.1).unwrap().contains(&(x + 1, y - 1)) &&
            letters.get(&l.2).unwrap().contains(&(x - 1, y + 1)) &&
            letters.get(&l.3).unwrap().contains(&(x + 1, y + 1))))
    .count()
}